room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/InvalidationTrackerBehavioralTest.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.room.integration.testapp.test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.room.Dao;
+import androidx.room.Database;
+import androidx.room.Entity;
+import androidx.room.Insert;
+import androidx.room.InvalidationTracker;
+import androidx.room.PrimaryKey;
+import androidx.room.Room;
+import androidx.room.RoomDatabase;
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+import androidx.test.filters.FlakyTest;
+import androidx.test.filters.LargeTest;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Regression test for a situation where an InvalidationTracker callback may intermittently be
+ * invoked too early, too late, or not at all, due to missing transactionality in tracking table
+ * code, when distinct database updates occur in close temporal proximity.
+ */
+@LargeTest
+@FlakyTest(
+        bugId = 154040286,
+        detail = "Behavioral test for potentially intermittent InvalidationTracker problems"
+)
+@RunWith(AndroidJUnit4.class)
+public class InvalidationTrackerBehavioralTest {
+    private ExecutorService mExecutorService;
+
+    @Before
+    public void setup() {
+        mExecutorService = Executors.newSingleThreadExecutor();
+    }
+
+    @After
+    public void tearDown() {
+        mExecutorService.shutdown();
+    }
+
+    @Test
+    public void testInserts() throws ExecutionException, InterruptedException {
+        testInserts(RoomDatabase.JournalMode.TRUNCATE);
+        // Use AUTOMATIC rather than WRITE_AHEAD_LOGGING to support running this test on older
+        // SDKs; effectively old SDKs will exercise the TRUNCATE case a second time this way, but
+        // who cares, it won't hurt
+        testInserts(RoomDatabase.JournalMode.AUTOMATIC);
```

Comment:
Lets have two different `@Test` and the one with `WRITE_AHEAD_LOGGING` should be annotated with `@SdkSuppress(minSdkVersion = Build.VERSION_CODES.JELLY_BEAN)`, that way we don't do double work on older APIs and the test with WAL will only run in APIs where WAL is available.